File: macos/Tests/CursorTokenCoreTests/AppUpdateTests.swift
import Foundation
import XCTest
@testable import CursorTokenCore

final class AppUpdateTests: XCTestCase {
    func testFixtureCases() throws {
        let root = try XCTUnwrap(try Fixtures.json("update_release_cases.json") as? [String: Any])

        for row in root["extract_sha"] as! [[String: Any]] {
            XCTAssertEqual(AppUpdate.extractSha(row["text"] as? String), row["sha"] as? String, row["name"] as? String ?? "?")
        }
        for row in root["same_sha"] as! [[String: Any]] {
            XCTAssertEqual(
                AppUpdate.sameSha(row["left"] as? String, row["right"] as? String),
                row["same"] as? Bool,
                "\(row["left"] ?? "") vs \(row["right"] ?? "")"
            )
        }
        for row in root["normalize_sha"] as! [[String: Any]] {
            XCTAssertEqual(AppUpdate.normalizeSha(row["input"] as? String), row["output"] as? String)
        }
        for row in root["allowed_urls"] as! [[String: Any]] {
            XCTAssertEqual(AppUpdate.isAllowedDownloadURL(row["url"] as? String), row["ok"] as? Bool, row["url"] as? String ?? "")
        }
        for row in root["packaged_windows"] as! [[String: Any]] {
            XCTAssertEqual(AppUpdate.looksPackagedWindows(row["path"] as? String), row["ok"] as? Bool, row["path"] as? String ?? "")
        }
        for row in root["packaged_macos"] as! [[String: Any]] {
            XCTAssertEqual(AppUpdate.looksPackagedMacos(row["path"] as? String), row["ok"] as? Bool, row["path"] as? String ?? "")
        }
        for row in root["parse_release"] as! [[String: Any]] {
            let parsed = AppUpdate.parseRelease(row["json"] as! [String: Any])
            let exp = row["expected"] as! [String: Any]
            XCTAssertEqual(parsed.tag, exp["tag"] as? String)
            if let version = exp["version"] as? String {
                XCTAssertEqual(parsed.version, version)
            }
            XCTAssertEqual(parsed.commitSha, exp["sha"] as? String)
            XCTAssertEqual(parsed.pageUrl, exp["page_url"] as? String)
            XCTAssertEqual(AppUpdate.findAsset(parsed, name: AppUpdate.windowsAssetName)?.url, exp["windows_url"] as? String)
            XCTAssertEqual(AppUpdate.findAsset(parsed, name: AppUpdate.macosAssetName)?.url, exp["macos_url"] as? String)
            XCTAssertEqual(AppUpdate.findAsset(parsed, name: AppUpdate.windowsAssetName)?.id, int64(exp["windows_asset_id"]))
            XCTAssertEqual(AppUpdate.findAsset(parsed, name: AppUpdate.macosAssetName)?.id, int64(exp["macos_asset_id"]))
        }
        for row in root["parse_release_page"] as! [[String: Any]] {
            let parsed = try AppUpdate.parseReleasePage(row["html"] as? String ?? "")
            XCTAssertEqual(parsed.commitSha, row["sha"] as? String, row["name"] as? String ?? "?")
            XCTAssertEqual(AppUpdate.findAsset(parsed, name: AppUpdate.windowsAssetName)?.url, row["windows_url"] as? String)
            XCTAssertEqual(AppUpdate.findAsset(parsed, name: AppUpdate.macosAssetName)?.url, row["macos_url"] as? String)
        }
        for row in root["parse_tag_ref"] as! [[String: Any]] {
            let data = try JSONSerialization.data(withJSONObject: row["json"] as Any)
            let json = String(data: data, encoding: .utf8)!
            XCTAssertEqual(AppUpdate.parseTagRefSha(json), row["sha"] as? String)
        }

        let sample = AppUpdate.parseRelease((root["parse_release"] as! [[String: Any]])[0]["json"] as! [String: Any])
        for row in root["evaluate"] as! [[String: Any]] {
            var release = sample
            if row["clear_release_sha"] as? Bool == true {
                release.commitSha = ""
            }
            if let tag = row["tag"] as? String {
                release.tag = tag
                if let version = row["version"] as? String { release.version = version }
            }
            let asset = (row["asset"] as? String) == "macos" ? AppUpdate.macosAssetName : AppUpdate.windowsAssetName
            let currentVersion = row["current_version"] as? String ?? AppUpdate.productVersion
            let decision = AppUpdate.evaluate(
                release: release,
                assetName: asset,
                currentSha: row["current_sha"] as? String ?? "",
                installedSha: row["installed_sha"] as? String ?? "",
                installedAssetId: int64(row["installed_asset_id"]),
                currentVersion: currentVersion
            )
            XCTAssertEqual(decision.available, row["available"] as? Bool, row["name"] as? String ?? "?")
            XCTAssertEqual(decision.upToDate, row["up_to_date"] as? Bool, row["name"] as? String ?? "?")
        }

        for row in root["extract_version"] as! [[String: Any]] {
            XCTAssertEqual(
                AppUpdate.productVersionFromRelease(row["tag"] as? String, body: row["body"] as? String),
                row["version"] as? String
            )
        }
        for row in root["compare_versions"] as! [[String: Any]] {
            XCTAssertEqual(
                AppUpdate.compareProductVersions(row["left"] as? String, row["right"] as? String),
                Int(int64(row["cmp"]))
            )
        }
        for row in root["choose_release"] as! [[String: Any]] {
            let officialTag = row["official_tag"] as? String ?? ""
            let official: AppRelease? = officialTag.isEmpty ? nil : AppRelease(
                tag: officialTag,
                version: row["official_version"] as? String ?? "",
                assets: sample.assets
            )
            let rolling = AppRelease(tag: row["rolling_tag"] as? String ?? "latest", assets: sample.assets)
            let chosen = AppUpdate.chooseRelease(
                official: official,
                rolling: rolling,
                currentVersion: row["current_version"] as? String
            )
            let choice = (official != nil && chosen?.tag == official?.tag) ? "official" : "rolling"
            XCTAssertEqual(choice, row["choice"] as? String, row["name"] as? String ?? "?")
        }
    }

    func testPrefersLightWindowsAssetThenFallsBackToFull() {
        let light = AppReleaseAsset(
            id: 9,
            name: AppUpdate.windowsLightAssetName,
            url: AppUpdate.assetDownloadURL(AppUpdate.windowsLightAssetName),
            size: 3_000_000
        )
        let full = AppReleaseAsset(
            id: 8,
            name: AppUpdate.windowsAssetName,
            url: AppUpdate.assetDownloadURL(AppUpdate.windowsAssetName),
            size: 60_000_000
        )
        let release = AppRelease(
            tag: "v2.1.0",
            version: "2.1.0",
            commitSha: "518192b000000000000000000000000000000000",
            assets: [light, full]
        )
        XCTAssertEqual(
            AppUpdate.assetNameCandidates(AppUpdate.windowsLightAssetName),
            [AppUpdate.windowsLightAssetName, AppUpdate.windowsAssetName]
        )
        XCTAssertEqual(
            AppUpdate.assetNameCandidates(AppUpdate.windowsAssetName),
            [AppUpdate.windowsAssetName]
        )
        XCTAssertEqual(AppUpdate.preferredWindowsAssetName(preferLight: true), AppUpdate.windowsLightAssetName)
        XCTAssertEqual(AppUpdate.preferredWindowsAssetName(preferLight: false), AppUpdate.windowsAssetName)
        XCTAssertEqual(AppUpdate.findPreferredAsset(release, name: AppUpdate.windowsLightAssetName)?.id, 9)
        XCTAssertEqual(AppUpdate.findPreferredAsset(release, name: AppUpdate.windowsAssetName)?.id, 8)

        let missingLight = AppRelease(
            tag: release.tag,
            version: release.version,
            commitSha: release.commitSha,
            assets: [full]
        )
        XCTAssertEqual(AppUpdate.findPreferredAsset(missingLight, name: AppUpdate.windowsLightAssetName)?.id, 8)

        let oversizedLight = AppRelease(
            tag: release.tag,
            version: release.version,
            commitSha: release.commitSha,
            assets: [
                AppReleaseAsset(
                    id: 11,
                    name: AppUpdate.windowsLightAssetName,
                    url: AppUpdate.assetDownloadURL(AppUpdate.windowsLightAssetName),
                    size: AppUpdate.maxLightZipBytes + 1
                ),
                full,
            ]
        )
        XCTAssertEqual(AppUpdate.findPreferredAsset(oversizedLight, name: AppUpdate.windowsLightAssetName)?.id, 8)

        let decision = AppUpdate.evaluate(
            release: release,
            assetName: AppUpdate.windowsLightAssetName,
            currentSha: "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
            currentVersion: "2.0.0"
        )
        XCTAssertTrue(decision.available)
        XCTAssertEqual(decision.asset?.id, 9)
    }

    func testIgnoresLegacyZipWhenCurrentAssetMissing() {
        let release = AppRelease(
            tag: "latest",
            commitSha: "518192b000000000000000000000000000000000",
            assets: [
                AppReleaseAsset(
                    id: 7,
                    name: AppUpdate.legacyWindowsAssetName,
                    url: AppUpdate.assetDownloadURL(AppUpdate.legacyWindowsAssetName)
                )
            ]
        )
        XCTAssertNil(AppUpdate.findPreferredAsset(release, name: AppUpdate.windowsAssetName))
        let decision = AppUpdate.evaluate(
            release: release,
            assetName: AppUpdate.windowsAssetName,
            currentSha: "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
        )
        XCTAssertFalse(decision.available)
        XCTAssertEqual(decision.message, "最新发布没有本平台安装包")
    }

    func testDisplayVersionAndInformationalSha() {
        XCTAssertEqual(AppUpdate.displayVersion(""), AppUpdate.productVersion)
        XCTAssertEqual(AppUpdate.displayVersion("518192b000000000000000000000000000000000"), "\(AppUpdate.productVersion) (518192b)")
        XCTAssertEqual(AppUpdate.shaFromInformationalVersion("\(AppUpdate.productVersion)+518192b"), "518192b")
        XCTAssertEqual(AppUpdate.shaFromInformationalVersion(AppUpdate.productVersion), "")
        XCTAssertEqual(AppUpdate.normalizeProductVersion("v\(AppUpdate.productVersion)"), AppUpdate.productVersion)
        XCTAssertTrue(AppUpdate.shouldFallbackFromApi(403))
        XCTAssertTrue(AppUpdate.shouldFallbackFromApi(429))
        XCTAssertFalse(AppUpdate.shouldFallbackFromApi(400))
    }

    func testShouldAutoCheckRespectsInterval() {
        let now = Date(timeIntervalSince1970: 1_784_000_000)
        XCTAssertTrue(AppUpdate.shouldAutoCheck(lastCheckAt: "", now: now, manual: false))
        XCTAssertTrue(AppUpdate.shouldAutoCheck(lastCheckAt: AppUpdate.nowIso(now.addingTimeInterval(-13 * 3600)), now: now, manual: false))
        XCTAssertFalse(AppUpdate.shouldAutoCheck(lastCheckAt: AppUpdate.nowIso(now.addingTimeInterval(-3600)), now: now, manual: false))
        XCTAssertTrue(AppUpdate.shouldAutoCheck(lastCheckAt: AppUpdate.nowIso(now.addingTimeInterval(-3600)), now: now, manual: true))
    }

    func testRememberedInstallAfterHelperKeepsOldShaOnFailure() {
        XCTAssertNil(AppUpdate.rememberedInstallAfterHelper(
            sha: "518192b000000000000000000000000000000000",
            assetId: 99,
            helperStarted: false
        ))
        let remembered = AppUpdate.rememberedInstallAfterHelper(
            sha: "518192B000000000000000000000000000000000",
            assetId: 99,
            helperStarted: true
        )
        XCTAssertEqual(remembered?.sha, "518192b000000000000000000000000000000000")
        XCTAssertEqual(remembered?.assetId, 99)
    }

    func testPendingInstallConfirmsOnlyAfterNewBinaryMatches() throws {
        let dir = FileManager.default.temporaryDirectory.appendingPathComponent("ctt-pending-\(UUID().uuidString)", isDirectory: true)
        try FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        defer { try? FileManager.default.removeItem(at: dir) }
        XCTAssertNil(AppUpdate.readPendingInstall(directory: dir))
        AppUpdate.writePendingInstall(directory: dir, sha: "0ad980084152a814bdaefeb3bb162184db07d64a", assetId: 7)
        let pending = try XCTUnwrap(AppUpdate.readPendingInstall(directory: dir))
        XCTAssertEqual(pending.sha, "0ad980084152a814bdaefeb3bb162184db07d64a")
        XCTAssertEqual(pending.assetId, 7)
        XCTAssertNil(AppUpdate.confirmPendingInstall(currentSha: "eb9a0a2a0512609588b3458d87443971eed07ec3", pending: pending))
        XCTAssertTrue(AppUpdate.pendingInstallFailed(currentSha: "eb9a0a2a0512609588b3458d87443971eed07ec3", pending: pending))
        XCTAssertFalse(AppUpdate.pendingInstallFailed(currentSha: "", pending: pending))
        let confirmed = AppUpdate.confirmPendingInstall(currentSha: "0ad9800", pending: pending)
        XCTAssertEqual(confirmed?.sha, pending.sha)
        AppUpdate.clearPendingInstall(directory: dir)
        XCTAssertNil(AppUpdate.readPendingInstall(directory: dir))
    }

    func testDetachedHelperCommandKeepsExternalVolumePathsQuoted() {
        let cmd = AppUpdate.detachedHelperLaunchCommand(
            script: "/tmp/apply.sh",
            pid: "12",
            source: "/tmp/src/CursorRemain.app",
            destination: "/Volumes/数据盘/Documents/Git/cursorremain/macos/dist/CursorRemain.app",
            expectedSha: "0ad9800",
            log: "/Users/me/Library/Logs/CursorRemain-update-helper.log"
        )
        XCTAssertTrue(cmd.hasPrefix("/usr/bin/nohup "))
        XCTAssertTrue(cmd.contains("'/Volumes/数据盘/Documents/Git/cursorremain/macos/dist/CursorRemain.app'"))
        XCTAssertTrue(cmd.hasSuffix(" &"))
        XCTAssertEqual(AppUpdate.shellQuote("it's"), "'it'\\''s'")
    }

    func testConfigRoundtripsAutoUpdateFields() throws {
        let dir = FileManager.default.temporaryDirectory.appendingPathComponent("ctt-update-cfg-\(UUID().uuidString)", isDirectory: true)
        try FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        defer { try? FileManager.default.removeItem(at: dir) }
        var cfg = AppConfig.default
        cfg.autoUpdateEnabled = false
        cfg.updateLastCheckAt = "2026-09-14T03:20:00Z"
        cfg.updateLastError = "x"
        cfg.updateInstalledSha = "518192b"
        cfg.updateInstalledAssetId = 101
        XCTAssertTrue(ConfigStore.save(cfg, to: dir))
        let loaded = ConfigStore.load(from: dir)
        XCTAssertFalse(loaded.autoUpdateEnabled)
        XCTAssertEqual(loaded.updateLastCheckAt, "2026-09-14T03:20:00Z")
        XCTAssertEqual(loaded.updateLastError, "x")
        XCTAssertEqual(loaded.updateInstalledSha, "518192b")
        XCTAssertEqual(loaded.updateInstalledAssetId, 101)
        cfg.colorMode = "dark"
        XCTAssertTrue(ConfigStore.save(cfg, to: dir))
        XCTAssertEqual(ConfigStore.load(from: dir).colorMode, "dark")
        let fresh = ConfigStore.normalize([:])
        XCTAssertTrue(fresh.autoUpdateEnabled)
        XCTAssertEqual(fresh.colorMode, "system")
        XCTAssertEqual(ConfigStore.normalize(["color_mode": "Blue"]).colorMode, "system")
    }

    private func int64(_ raw: Any?) -> Int64 {
        if let n = raw as? NSNumber { return n.int64Value }
        if let i = raw as? Int { return Int64(i) }
        if let s = raw as? String, let v = Int64(s) { return v }
        return 0
    }
}

File: windows/CursorTokenCore.Tests/AppUpdateTests.cs
using System.IO.Compression;
using System.Text.Json;
using CursorTokenCore;
using Xunit;

namespace CursorTokenCore.Tests;

public class AppUpdateTests
{
    static JsonElement Load()
    {
        var dir = new DirectoryInfo(AppContext.BaseDirectory);
        for (var i = 0; i < 12 && dir is not null; i++)
        {
            var candidate = Path.Combine(dir.FullName, "fixtures", "update_release_cases.json");
            if (File.Exists(candidate)) return JsonDocument.Parse(File.ReadAllText(candidate)).RootElement.Clone();
            dir = dir.Parent;
        }
        var cwd = new DirectoryInfo(Directory.GetCurrentDirectory());
        for (var i = 0; i < 12 && cwd is not null; i++)
        {
            var candidate = Path.Combine(cwd.FullName, "fixtures", "update_release_cases.json");
            if (File.Exists(candidate)) return JsonDocument.Parse(File.ReadAllText(candidate)).RootElement.Clone();
            cwd = cwd.Parent;
        }
        throw new FileNotFoundException("update_release_cases.json");
    }

    [Fact]
    public void FixtureCases()
    {
        var root = Load();
        foreach (var row in root.GetProperty("extract_sha").EnumerateArray())
            Assert.Equal(row.GetProperty("sha").GetString(), AppUpdate.ExtractSha(row.GetProperty("text").GetString()));

        foreach (var row in root.GetProperty("same_sha").EnumerateArray())
            Assert.Equal(row.GetProperty("same").GetBoolean(), AppUpdate.SameSha(row.GetProperty("left").GetString(), row.GetProperty("right").GetString()));

        foreach (var row in root.GetProperty("normalize_sha").EnumerateArray())
            Assert.Equal(row.GetProperty("output").GetString(), AppUpdate.NormalizeSha(row.GetProperty("input").GetString()));

        foreach (var row in root.GetProperty("allowed_urls").EnumerateArray())
            Assert.Equal(row.GetProperty("ok").GetBoolean(), AppUpdate.IsAllowedDownloadUrl(row.GetProperty("url").GetString()));

        foreach (var row in root.GetProperty("packaged_windows").EnumerateArray())
            Assert.Equal(row.GetProperty("ok").GetBoolean(), AppUpdate.LooksPackagedWindows(row.GetProperty("path").GetString()));

        foreach (var row in root.GetProperty("packaged_macos").EnumerateArray())
            Assert.Equal(row.GetProperty("ok").GetBoolean(), AppUpdate.LooksPackagedMacos(row.GetProperty("path").GetString()));

        foreach (var row in root.GetProperty("parse_release").EnumerateArray())
        {
            var parsed = AppUpdate.ParseRelease(row.GetProperty("json").GetRawText());
            var exp = row.GetProperty("expected");
            Assert.Equal(exp.GetProperty("tag").GetString(), parsed.Tag);
            if (exp.TryGetProperty("version", out var versionEl))
                Assert.Equal(versionEl.GetString(), parsed.Version);
            Assert.Equal(exp.GetProperty("sha").GetString(), parsed.CommitSha);
            Assert.Equal(exp.GetProperty("page_url").GetString(), parsed.PageUrl);
            Assert.Equal(exp.GetProperty("windows_url").GetString(), AppUpdate.FindAsset(parsed, AppUpdate.WindowsAssetName)?.Url);
            Assert.Equal(exp.GetProperty("macos_url").GetString(), AppUpdate.FindAsset(parsed, AppUpdate.MacosAssetName)?.Url);
            Assert.Equal(exp.GetProperty("windows_asset_id").GetInt64(), AppUpdate.FindAsset(parsed, AppUpdate.WindowsAssetName)?.Id);
            Assert.Equal(exp.GetProperty("macos_asset_id").GetInt64(), AppUpdate.FindAsset(parsed, AppUpdate.MacosAssetName)?.Id);
        }

        foreach (var row in root.GetProperty("parse_release_page").EnumerateArray())
        {
            var parsed = AppUpdate.ParseReleasePage(row.GetProperty("html").GetString() ?? "");
            Assert.Equal(row.GetProperty("sha").GetString(), parsed.CommitSha);
            Assert.Equal(row.GetProperty("windows_url").GetString(), AppUpdate.FindAsset(parsed, AppUpdate.WindowsAssetName)?.Url);
            Assert.Equal(row.GetProperty("macos_url").GetString(), AppUpdate.FindAsset(parsed, AppUpdate.MacosAssetName)?.Url);
        }

        foreach (var row in root.GetProperty("parse_tag_ref").EnumerateArray())
            Assert.Equal(row.GetProperty("sha").GetString(), AppUpdate.ParseTagRefSha(row.GetProperty("json").GetRawText()));

        var sample = AppUpdate.ParseRelease(root.GetProperty("parse_release")[0].GetProperty("json").GetRawText());
        foreach (var row in root.GetProperty("evaluate").EnumerateArray())
        {
            var release = sample;
            if (row.TryGetProperty("clear_release_sha", out var clear) && clear.GetBoolean())
            {
                release = new AppRelease
                {
                    Tag = sample.Tag,
                    CommitSha = "",
                    PageUrl = sample.PageUrl,
                    PublishedAt = sample.PublishedAt,
                    Assets = sample.Assets,
                };
            }
            if (row.TryGetProperty("tag", out var tagEl))
            {
                release = new AppRelease
                {
                    Tag = tagEl.GetString() ?? release.Tag,
                    Version = row.TryGetProperty("version", out var verEl) ? verEl.GetString() ?? "" : release.Version,
                    Body = release.Body,
                    CommitSha = release.CommitSha,
                    PageUrl = release.PageUrl,
                    PublishedAt = release.PublishedAt,
                    Assets = release.Assets,
                };
            }
            var asset = row.GetProperty("asset").GetString() == "macos" ? AppUpdate.MacosAssetName : AppUpdate.WindowsAssetName;
            var currentVersion = row.TryGetProperty("current_version", out var curVer)
                ? curVer.GetString()
                : AppUpdate.ProductVersion;
            var decision = AppUpdate.Evaluate(
                release,
                asset,
                row.GetProperty("current_sha").GetString() ?? "",
                row.GetProperty("installed_sha").GetString() ?? "",
                row.GetProperty("installed_asset_id").GetInt64(),
                currentVersion);
            Assert.Equal(row.GetProperty("available").GetBoolean(), decision.Available);
            Assert.Equal(row.GetProperty("up_to_date").GetBoolean(), decision.UpToDate);
        }

        foreach (var row in root.GetProperty("extract_version").EnumerateArray())
            Assert.Equal(row.GetProperty("version").GetString(), AppUpdate.ProductVersionFromRelease(row.GetProperty("tag").GetString(), row.GetProperty("body").GetString()));

        foreach (var row in root.GetProperty("compare_versions").EnumerateArray())
            Assert.Equal(row.GetProperty("cmp").GetInt32(), AppUpdate.CompareProductVersions(row.GetProperty("left").GetString(), row.GetProperty("right").GetString()));

        foreach (var row in root.GetProperty("choose_release").EnumerateArray())
        {
            AppRelease? official = null;
            var officialTag = row.GetProperty("official_tag").GetString() ?? "";
            if (officialTag.Length > 0)
            {
                official = new AppRelease
                {
                    Tag = officialTag,
                    Version = row.GetProperty("official_version").GetString() ?? "",
                    Assets = sample.Assets,
                };
            }
            var rolling = new AppRelease { Tag = row.GetProperty("rolling_tag").GetString() ?? "latest", Assets = sample.Assets };
            var chosen = AppUpdate.ChooseRelease(official, rolling, row.GetProperty("current_version").GetString());
            Assert.Equal(row.GetProperty("choice").GetString(), chosen?.Tag == official?.Tag && official is not null ? "official" : "rolling");
        }
    }

    [Fact]
    public void PrefersLightWindowsAssetThenFallsBackToFull()
    {
        var light = new AppReleaseAsset
        {
            Name = AppUpdate.WindowsLightAssetName,
            Url = AppUpdate.AssetDownloadUrl(AppUpdate.WindowsLightAssetName),
            Id = 9,
            Size = 3_000_000,
        };
        var full = new AppReleaseAsset
        {
            Name = AppUpdate.WindowsAssetName,
            Url = AppUpdate.AssetDownloadUrl(AppUpdate.WindowsAssetName),
            Id = 8,
            Size = 60_000_000,
        };
        var release = new AppRelease
        {
            Tag = "v2.1.0",
            Version = "2.1.0",
            CommitSha = "518192b000000000000000000000000000000000",
            Assets = [light, full],
        };
        Assert.Equal(
            new[] { AppUpdate.WindowsLightAssetName, AppUpdate.WindowsAssetName },
            AppUpdate.AssetNameCandidates(AppUpdate.WindowsLightAssetName));
        Assert.Equal(
            new[] { AppUpdate.WindowsAssetName },
            AppUpdate.AssetNameCandidates(AppUpdate.WindowsAssetName));
        Assert.Equal(AppUpdate.WindowsLightAssetName, AppUpdate.PreferredWindowsAssetName(true));
        Assert.Equal(AppUpdate.WindowsAssetName, AppUpdate.PreferredWindowsAssetName(false));

        var lightPick = AppUpdate.FindPreferredAsset(release, AppUpdate.WindowsLightAssetName);
        Assert.Equal(9, lightPick?.Id);
        var fullPick = AppUpdate.FindPreferredAsset(release, AppUpdate.WindowsAssetName);
        Assert.Equal(8, fullPick?.Id);

        var missingLight = new AppRelease
        {
            Tag = release.Tag,
            Version = release.Version,
            CommitSha = release.CommitSha,
            Assets = [full],
        };
        Assert.Equal(8, AppUpdate.FindPreferredAsset(missingLight, AppUpdate.WindowsLightAssetName)?.Id);

        var oversizedLight = new AppRelease
        {
            Tag = release.Tag,
            Version = release.Version,
            CommitSha = release.CommitSha,
            Assets =
            [
                new()
                {
                    Name = AppUpdate.WindowsLightAssetName,
                    Url = AppUpdate.AssetDownloadUrl(AppUpdate.WindowsLightAssetName),
                    Id = 11,
                    Size = AppUpdate.MaxLightZipBytes + 1,
                },
                full,
            ],
        };
        Assert.Equal(8, AppUpdate.FindPreferredAsset(oversizedLight, AppUpdate.WindowsLightAssetName)?.Id);

        var decision = AppUpdate.Evaluate(
            release,
            AppUpdate.WindowsLightAssetName,
            "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
            currentVersion: "2.0.0");
        Assert.True(decision.Available);
        Assert.Equal(9, decision.Asset?.Id);
    }

    [Fact]
    public void DetectsDesktopRuntimeAndSharedFramework()
    {
        Assert.True(AppUpdate.RunsOnSharedFramework(
            @"C:\Program Files\dotnet\shared\Microsoft.NETCore.App\8.0.20\System.Private.CoreLib.dll"));
        Assert.True(AppUpdate.RunsOnSharedFramework(
            "/usr/share/dotnet/shared/Microsoft.WindowsDesktop.App/8.0.20/System.Windows.Forms.dll"));
        Assert.False(AppUpdate.RunsOnSharedFramework(""));
        Assert.False(AppUpdate.RunsOnSharedFramework(@"C:\Users\me\AppData\Local\Temp\.net\CursorRemain\coreclr.dll"));

        var root = Path.Combine(Path.GetTempPath(), "ctt-fx-" + Guid.NewGuid().ToString("N"));
        var runtime = Path.Combine(root, "8.0.20");
        Directory.CreateDirectory(runtime);
        try
        {
            File.WriteAllText(Path.Combine(runtime, "System.Windows.Forms.dll"), "fx");
            Assert.True(AppUpdate.ContainsDesktopRuntime8(root));
            Assert.True(AppUpdate.HasWindowsDesktopRuntime([root]));
            Assert.False(AppUpdate.HasWindowsDesktopRuntime([Path.Combine(root, "missing")]));
            Assert.True(AppUpdate.CanUseWindowsLightUpdate(hasDesktopRuntime: true, runsOnSharedFramework: false));
            Assert.False(AppUpdate.CanUseWindowsLightUpdate(hasDesktopRuntime: false, runsOnSharedFramework: false));
            Assert.True(AppUpdate.CanUseWindowsLightUpdate(hasDesktopRuntime: false, runsOnSharedFramework: true));
        }
        finally
        {
            try { Directory.Delete(root, true); } catch { }
        }
    }

    [Fact]
    public void IgnoresLegacyZipWhenCurrentAssetMissing()
    {
        var release = new AppRelease
        {
            Tag = "latest",
            CommitSha = "518192b000000000000000000000000000000000",
            Assets =
            [
                new()
                {
                    Name = AppUpdate.LegacyWindowsAssetName,
                    Url = AppUpdate.AssetDownloadUrl(AppUpdate.LegacyWindowsAssetName),
                    Id = 7,
                },
            ],
        };
        Assert.Null(AppUpdate.FindPreferredAsset(release, AppUpdate.WindowsAssetName));
        var decision = AppUpdate.Evaluate(release, AppUpdate.WindowsAssetName, "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa");
        Assert.False(decision.Available);
        Assert.Equal("最新发布没有本平台安装包", decision.Message);
    }

    [Fact]
    public void DisplayVersionAndInformationalSha()
    {
        var version = AppUpdate.ProductVersion;
        Assert.False(string.IsNullOrWhiteSpace(version));
        Assert.Equal(version, AppUpdate.DisplayVersion(""));
        Assert.Equal($"{version} (518192b)", AppUpdate.DisplayVersion("518192b000000000000000000000000000000000"));
        Assert.Equal("518192b", AppUpdate.ShaFromInformationalVersion(version + "+518192b"));
        Assert.Equal("", AppUpdate.ShaFromInformationalVersion(version));
        Assert.Equal(version, AppUpdate.NormalizeProductVersion("v" + version));
        Assert.True(AppUpdate.CompareProductVersions(version, version) == 0);
        Assert.True(AppUpdate.ShouldFallbackFromApi(403));
        Assert.True(AppUpdate.ShouldFallbackFromApi(429));
        Assert.False(AppUpdate.ShouldFallbackFromApi(400));

        var dir = new DirectoryInfo(AppContext.BaseDirectory);
        for (var i = 0; i < 12 && dir is not null; i++)
        {
            var candidate = Path.Combine(dir.FullName, "VERSION");
            if (File.Exists(candidate))
            {
                Assert.Equal(File.ReadAllText(candidate).Trim(), version);
                return;
            }
            dir = dir.Parent;
        }
        throw new FileNotFoundException("VERSION");
    }

    [Fact]
    public void RememberedInstallAfterHelperKeepsOldShaOnFailure()
    {
        Assert.Null(AppUpdate.RememberedInstallAfterHelper("518192b000000000000000000000000000000000", 99, false));
        var remembered = AppUpdate.RememberedInstallAfterHelper("518192B000000000000000000000000000000000", 99, true);
        Assert.NotNull(remembered);
        Assert.Equal("518192b000000000000000000000000000000000", remembered.Value.Sha);
        Assert.Equal(99, remembered.Value.AssetId);
    }

    [Fact]
    public void PendingInstallConfirmsOnlyAfterNewBinaryMatches()
    {
        var dir = Path.Combine(Path.GetTempPath(), "ctt-pending-" + Guid.NewGuid().ToString("N"));
        Directory.CreateDirectory(dir);
        try
        {
            Assert.Null(AppUpdate.ReadPendingInstall(dir));
            AppUpdate.WritePendingInstall(dir, "0ad980084152a814bdaefeb3bb162184db07d64a", 7);
            var pending = AppUpdate.ReadPendingInstall(dir);
            Assert.NotNull(pending);
            Assert.Equal("0ad980084152a814bdaefeb3bb162184db07d64a", pending.Value.Sha);
            Assert.Equal(7, pending.Value.AssetId);
            Assert.Null(AppUpdate.ConfirmPendingInstall("eb9a0a2a0512609588b3458d87443971eed07ec3", pending));
            Assert.True(AppUpdate.PendingInstallFailed("eb9a0a2a0512609588b3458d87443971eed07ec3", pending));
            Assert.False(AppUpdate.PendingInstallFailed("", pending));
            var confirmed = AppUpdate.ConfirmPendingInstall("0ad9800", pending);
            Assert.NotNull(confirmed);
            Assert.Equal(pending.Value.Sha, confirmed.Value.Sha);
            AppUpdate.ClearPendingInstall(dir);
            Assert.Null(AppUpdate.ReadPendingInstall(dir));
        }
        finally
        {
            try { Directory.Delete(dir, true); } catch { }
        }
    }

    [Fact]
    public void ShouldAutoCheckRespectsInterval()
    {
        var now = new DateTimeOffset(2026, 9, 14, 12, 0, 0, TimeSpan.Zero);
        Assert.True(AppUpdate.ShouldAutoCheck("", now, false));
        Assert.True(AppUpdate.ShouldAutoCheck(AppUpdate.NowIso(now.AddHours(-13)), now, false));
        Assert.False(AppUpdate.ShouldAutoCheck(AppUpdate.NowIso(now.AddHours(-1)), now, false));
        Assert.True(AppUpdate.ShouldAutoCheck(AppUpdate.NowIso(now.AddHours(-1)), now, true));
    }

    [Fact]
    public void ExtractZipRejectsPathEscapeAndFindsExe()
    {
        var root = Path.Combine(Path.GetTempPath(), "ctt-update-test-" + Guid.NewGuid().ToString("N"));
        Directory.CreateDirectory(root);
        try
        {
            var zip = Path.Combine(root, "ok.zip");
            using (var archive = ZipFile.Open(zip, ZipArchiveMode.Create))
            {
                var entry = archive.CreateEntry("CursorRemain.exe");
                using var w = new StreamWriter(entry.Open());
                w.Write("exe");
            }
            var dest = Path.Combine(root, "out");
            AppUpdate.ExtractZipSafe(zip, dest);
            Assert.Equal(Path.Combine(dest, "CursorRemain.exe"), AppUpdate.FindExtractedWindowsExe(dest));

            var legacyZip = Path.Combine(root, "legacy.zip");
            using (var archive = ZipFile.Open(legacyZip, ZipArchiveMode.Create))
            {
                var entry = archive.CreateEntry("CursorTokenTray.exe");
                using var w = new StreamWriter(entry.Open());
                w.Write("old");
            }
            var legacyDest = Path.Combine(root, "legacy-out");
            AppUpdate.ExtractZipSafe(legacyZip, legacyDest);
            Assert.Equal(Path.Combine(legacyDest, "CursorTokenTray.exe"), AppUpdate.FindExtractedWindowsExe(legacyDest));

            var evil = Path.Combine(root, "evil.zip");
            using (var archive = ZipFile.Open(evil, ZipArchiveMode.Create))
                archive.CreateEntry("../escape.exe");
            Assert.Throws<InvalidOperationException>(() => AppUpdate.ExtractZipSafe(evil, Path.Combine(root, "evil-out")));
        }
        finally
        {
            try { Directory.Delete(root, true); } catch { }
        }
    }

    [Fact]
    public void ConfigRoundtripsAutoUpdateFields()
    {
        var dir = Path.Combine(Path.GetTempPath(), "ctt-update-cfg-" + Guid.NewGuid().ToString("N"));
        Directory.CreateDirectory(dir);
        try
        {
            var cfg = new AppConfig
            {
                AutoUpdateEnabled = false,
                UpdateLastCheckAt = "2026-09-14T03:20:00Z",
                UpdateLastError = "x",
                UpdateInstalledSha = "518192b",
                UpdateInstalledAssetId = 101,
            };
            ConfigStore.Save(cfg, dir);
            var loaded = ConfigStore.Load(dir);
            Assert.False(loaded.AutoUpdateEnabled);
            Assert.Equal("2026-09-14T03:20:00Z", loaded.UpdateLastCheckAt);
            Assert.Equal("x", loaded.UpdateLastError);
            Assert.Equal("518192b", loaded.UpdateInstalledSha);
            Assert.Equal(101, loaded.UpdateInstalledAssetId);

            var fresh = ConfigStore.Normalize(JsonDocument.Parse("{}").RootElement);
            Assert.True(fresh.AutoUpdateEnabled);
        }
        finally
        {
            try { Directory.Delete(dir, true); } catch { }
        }
    }
}
